contrib/hualloc/hu_alloc.h

```diff
@@ -0,0 +1,1615 @@
+    std::atomic<ui64> &mask = memFreeBits[block * BITS_OFFSET_MULT + chunk];
+    Y_ASSERT((mask & allocMask) == 0);
+    ui64 newMask = mask.fetch_add(allocMask) + allocMask;
+    if (_mm_popcnt_u64(newMask) == 16) { // if there is non trivial number of free allocations reuse the block
+        // can modify CanAllocBlockMask for the current segment only
+        if (segment == pThreadSegment) {
+            TSegmentSizeContext &sCtx = segHdr->SizeCtx[pow2];
+            sCtx.CanAllocBlockMask |= (1ull << block);
+        } else {
+            segHdr->RetestBlockMask |= (1ull << block); // no need to acquire lock at this point, the flag will be used later
+        }
+    }
+    if (newMask == ALL_BITS) {
+        TGlobalSizeContext &gCtx = GlobalSizeCtx[pow2];
+        std::atomic<ui64> *blockFreeBits = memFreeBits + BITS_OFFSET_MULT;
+        ui64 newFreeChunkCount = blockFreeBits[block].fetch_add(1) + 1;
+        if (newFreeChunkCount == (ui64)gCtx.ChunkCount) {
+            TSegmentSizeContext &sCtx = segHdr->SizeCtx[pow2];
+            BlockFree(segment, block, &sCtx);
+        }
+    }
+}
+
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////
+// common functions for all segment* allocations
+static bool ProcessRetestMaskLocked(char *segment)
+{
+    TSegmentHeader *segHdr = (TSegmentHeader*)segment;
+    ui64 retestMask = segHdr->RetestBlockMask.exchange(0);
+    if (retestMask == 0) {
+        return false;
+    }
+    unsigned long block;
+    while (_BitScanForward64(&block, retestMask)) {
+        retestMask -= 1ull << block;
+        ui8 pow2 = segHdr->CurAllocLevel[block];
+        if (pow2 < MIN_POW2) {
+            ; // this can happen, retest mask does not guarantee there is something to be done
+        } else if (pow2 < NEED_2LAYER) {
+            ProcessRetestTinySizeLocked(segment, pow2, block);
+        } else if (pow2 < BLOCK_SIZE_LN) {
+            ProcessRetestSmallSizeLocked(segment, pow2, block);
+        } else {
+            ; // this can happen, retest mask does not guarantee there is something to be done
+        }
+    }
+    return true;
+}
+
+
+static void DumpLocalAllocMasksLocked(char *segment)
+{
+    DumpTinyLocalAllocMasksLocked(segment);
+    DumpSmallLocalAllocMasksLocked(segment);
+}
+
+
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////
+static yint ReclaimKeepSize = 1024 * (1ull << 20);
+static yint ReclaimMaxReclaim = 512 * (1ull << 20);
+static THREAD_FUNC_RETURN ReclaimThread(void*)
+{
+    // keep & max can be separate for large & segment spaces
+    for (;;) {
+        Sleep(1000);
+        LargeReclaim(ReclaimKeepSize, ReclaimMaxReclaim);
+        SegmentReclaim(ReclaimKeepSize, ReclaimMaxReclaim);
+    }
+}
+
+void RunHuReclaim()
+{
+#ifdef _win_
+    CreateThread(0, 0, ReclaimThread, 0, 0, 0);
+#else
+    pthread_t tid;
+    pthread_create(&tid, 0, ReclaimThread, 0);
+#endif
```

Comment:
Once we employed C++11 atomics, we can use portable C++11 features for threads, sleep, yield and TLS.

It's especially important for [Win32 CreateThread](https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createthread) that shouldn't be used in C++ apps since it skips CRT-specific initialization:

> A thread in an executable that calls the C run-time library (CRT) should use the _beginthreadex and _endthreadex functions for thread management rather than CreateThread and ExitThread; this requires the use of the multithreaded version of the CRT. If a thread created using CreateThread calls the CRT, the CRT may terminate the process in low-memory conditions.

Reply:
That's excellent that CreateThread() skips CRT.  We need no CRT features here and do not want to interfere with CRT allocating thread specific buffers etc for this internal thread.